### ISSUE TITLE
ER: Filters Sidebar shifts to the Center When there are No Project Cards on "Projects" and "Homepage"

### ISSUE THREAD

**Issue by agosmou** [hackforla/website#4769]
### Emergent Requirement - Problem
On the [Projects](https://www.hackforla.org/projects/) and [Homepage](https://www.hackforla.org), when you select an assortment of project filters so that there is no project that meets those specifications, the filters sidebar will shift towards the center. See this comment: https://github.com/hackforla/website/pull/4760#pullrequestreview-1454823740

### Issue you discovered this emergent requirement in
- #4753

### Date discovered 
- June 1st, 2023

### Did you have to do something temporarily
- [ ] YES
- [x] NO

### Who was involved
@agosmou @t-will-gillis  @roslynwythe 

### What happens if this is not addressed
- people using the filters are confused

### Recommended Action Items
- [X] Make a new issue
- [x] Discuss with team
- [x] Let a Team Lead know

### Potential solutions [draft]
I would probably suggest the assignee to find a way to make sure the container that holds all the cards does not disappear when there are no cards.

**Comment by ExperimentsInHonesty** [comment]
I recommend that it throws a message in place of the cards, when there are no projects to show.  The message should be some helpful message about how to use the search.  See existing websites how they do that, when there are no results for a search.

**Comment by Adastros** [comment]
- A solution for this was submitted for the toolkit page filter implementation in pull request #4764 for issue #3985.  See screenshot below. 

The code for it can be transferred over to the projects filter. 

<details>
<summary>Toolkit Page Filter - No Results</summary>

![image](https://github.com/hackforla/website/assets/48114590/0600e866-6839-434e-999a-b877db7fd4be)

</details>

**Comment by ExperimentsInHonesty** [comment]
@Adastros Thank you for highlight the work already done.  We should 100% reuse that solution.  Can we reuse the code by referencing it, or do we have to do something to make it reusable first.  It would be less than optimal if we just copied and repeated it.  

Danny are you planning on writing the issue for this?  If not, that's fine, just want to know, so that if you don't want to, we can let other people know that they could write it.

**Comment by Retr0-XD** [comment]
Nice

**Comment by Adastros** [comment]
Hi @ExperimentsInHonesty, we won't be able to reuse the code by referencing it. The code will need to updated to make it reusable. 

Sorry, the intention of my message was to document/ notify the team of the potential solution since it was submitted with pull request #4764. 

I don't think I'll have the capacity to write the issue for this ER this week, but I'll try to provide as much info in this ER to help with the issue writing process. 

I plan to do the following

- ask the dev that worked on the solution to 
  - leave a comment in this ER to explain and document the code for the solution to help with the issue creation process and the dev that will be working on the issue created from this ER
  - leave the solution in pull request #4764 as-is. The issue that will be created for this ER can reference that code while refactoring it to make it reusable on any filter component.

**Comment by Adastros** [comment]
One more thing, the issue for this ER may need to add issue #3985 from pull request #4764 as a dependency so that the project and toolkit filters are implemented with the same functionality

**Comment by kwangric** [comment]
# How to Implement the No Results Message on Another Page
- The "No Results" feature is based off the design in Figma. On the toolkit page, whenever no cards were displayed, I called a function that inserts the HTML for the message with the current applied filters listed. When cards were displayed again, the message is cleared.
## Files to reference
- [Figma:](https://www.figma.com/file/0RRPy1Ph7HafI3qOITg0Mr/Hack-for-LA-Website?type=design&node-id=24097-286538) Contains the final design for the No Results message. Note that I did not include the message about checking the spelling since that applies to the search bar.
- [toolkit.js:](https://github.com/hackforla/website/blob/c536e346051b60957d46c3bcd62eef9118314645/assets/js/toolkit.js) This is an example of the code to display the "No results" message, including the HTML.
- [current-guides.html:](https://github.com/hackforla/website/blob/c536e346051b60957d46c3bcd62eef9118314645/_includes/current-guides.html) This is an example of how the HTML page is set up to allow the "No results" message to be inserted (see line 55)
## Recommended workflow
- Begin with reviewing the design on the projects page on [Figma](https://www.figma.com/file/0RRPy1Ph7HafI3qOITg0Mr/Hack-for-LA-Website?type=design&node-id=24097-286538)
- Review [toolkit.js:](https://github.com/hackforla/website/blob/c536e346051b60957d46c3bcd62eef9118314645/assets/js/toolkit.js), specifically the functions that display the message (see functions below)
- Customize the functions as needed, depending on how your filters are set up.
## Functions that will need to be changes to reflect you page's needs:
- noResultsMessageComponent (line 520 in `toolkit.js`) takes the current page filters as a parameter and returns HTML with all the applied filters listed as per the FIgma design. This will probably need to be customized to match the filter categories needed.
- displayNoResultsMessage (line 505 in `toolkit.js`) inserts the HTML that displays the "No results" message where the class `no-results-message` is. This function also takes the current page filters as a parameter and calls the function above.
- You will need a way to check and call displayNoResultsMessage if no data is displayed (line 247-251 in `toolkit.js`).

**Comment by chrismenke45** [comment]
Issue ETA: Friday 6/30
Availability: Weekdays 2-4pm

**Comment by chrismenke45** [comment]
Issue Created #4903 (still draft)
Updated ETA: tomorrow (7/1) EOD

**Comment by chrismenke45** [comment]
Issue #4903 ready for review

**Comment by ExperimentsInHonesty** [comment]
- on issue #4903 I left the following comment
It seems like this issue is to copy the code from the toolkit page and make some changes to it.  

What we should be doing is making an issue to refactor the code so that it can work for the toolkit page (and any other page) by calling a file that has the specific page variables in it. (filter names, messages, etc.).  

Once this refactor is done, then this issue can be revised to setup the file for this page with its variables in it.  And then call the script with the variable file to deliver a customized result to this page.

**Comment by chrismenke45** [comment]
Updated ETA: 7/5 EOD

**Comment by chrismenke45** [comment]
New issue added #4769. Both issues still drafts. Updated ETA: 7/6 EOD

**Comment by chrismenke45** [comment]
I am taking a break from HFLA and will not be able to complete this issue.

**Comment by Adastros** [comment]
Availability (PST): 10 PM - 12 AM 18JUL2023. 19JUL2023
ETA (PST): End of Day 19JUL2023

**Comment by kwangric** [comment]
@Adastros While helping Niki with parts of issue #4136, I noticed that she was adding code that will include the "No Results Message" for the projects page. I'm assuming this is because it was included in the Figma design for the issue, similarly to when I added it to the toolkit page when I was adding the dropdown filters.

Should we wait until we finish refactoring the code for the message, as per Bonnie's last comment, before adding it to the projects page?

**Comment by Adastros** [comment]
Hi @kwangric, thanks for catching that! It's outside the scope of issue #4136, but it should be okay to keep if it was already fully implemented to avoid undoing all the work that was put into adding the feature. We just need to make sure to mention in the pull request that the feature was implemented and reference this ER issue directly. This ER will also need to reference the pull request too.

If it hasn't been fully implemented, I would not include it the pull request to stick to the intended scope of the issue and simplify traceability of the feature. 

Please let me know if it was fully implemented or not as soon as possible since this issue is time sensitive. I'll put a hold on this ER until then.

**Comment by kwangric** [comment]
@Adastros Got an update today that the feature hasn't been fully implemented and that we can move forward with creating a new issue for it.

**Comment by Adastros** [comment]
Thanks @kwangric! Issue #4903 is ready for prioritization.

**Comment by Adastros** [comment]
Closing this ER as issue #4903 is completed and closed

**Comment by ExperimentsInHonesty** [comment]
- According to the notes in 4912.  #4912 was closed as unplanned as #4903 covered it.
